### PR TITLE
Explicit return generation and error handling

### LINKED ISSUE #38: error: expected instruction opcode  . Error received  when llvm code was passed to  ```bash ../brilisp/run.sh``` when trying to compile a matrix multiplication  c-lisp code
When the output from  `$ guile ../../utils/sexp-json.scm < matrix_multiplication.sexp | python3 ../../c-lisp.py | python3 ../../brilisp.py | python ../../llvm.py` was fed into  `bash ../brilisp/run.sh` , the below shown error was thrown.
```
$ guile ../../utils/sexp-json.scm < matrix_multiplication.sexp | python3 ../../c-lisp.py | python3 ../../brilisp.py | python ../../llvm.py| bash ../brilisp/run.sh
/tmp/tmp.SxEdPKa4GH.ll:167:1: error: expected instruction opcode
  167 | }
      | ^
1 error generated.
../brilisp/run.sh: line 9: /tmp/tmp.M2Q2136kah.out: Permission denied 
```

This is the code that was written in c-lisp (matrix_multiplication.sexp) which was given as an input to  guile ../../utils/sexp-json.scm
```
(c-lisp
	(define ((matrix_multiplication void) (xout (ptr float)) (x (ptr float)) (w (ptr float)) (n int) (d int)) 
		(declare (i int))
		(for ( (set i 0) (lt i d) (set i (add i 1)) )
			(declare (val float))
			(set val 0.0)
			(declare (j int))
			(for ((set j 0) (lt j n) (set j (add j 1)))
			   (set val (fadd val (fmul (load (ptradd w (add (mul i n) j))) (load (ptradd x j)))))
			)
			(store (ptradd xout i) val)
		) 
	)
)
```
This was the input given to `bash  ../brilisp/run.sh`

```
$ guile ../../utils/sexp-json.scm < matrix_multiplication.sexp | python3 ../../c-lisp.py | python3 ../../brilisp.py | python ../../llvm.py
; ModuleID = ""
target triple = "unknown-unknown-unknown"
target datalayout = ""

define void @"matrix_multiplication"(float* %"xout", float* %"x", float* %"w", i32 %"n", i32 %"d")
{
alloca-hwdeyruq:
  %"xout.1" = alloca float*
  %"x.1" = alloca float*
  %"w.1" = alloca float*
  %"n.1" = alloca i32
  %"d.1" = alloca i32
  %"tmp_clisp-myvvztwi" = alloca i32
  %"i" = alloca i32
  %"tmp_clisp.inp_0-tifpaiid" = alloca i32
  %"tmp_clisp.inp_1-qqagurna" = alloca i32
  %"tmp_clisp.cond-yoeqgrzo" = alloca i1
  %"tmp_clisp-yzjtqbcz" = alloca float
  %"val.twcwjsjw" = alloca float
  %"tmp_clisp-eqmzlcai" = alloca i32
  %"j.twcwjsjw" = alloca i32
  %"tmp_clisp.inp_0-lgfovtpl" = alloca i32
  %"tmp_clisp.inp_1-bguarjjv" = alloca i32
  %"tmp_clisp.cond-prqqdmwu" = alloca i1
  %"tmp_clisp.inp_0-xvbgbjas" = alloca float
  %"tmp_clisp.inp_0-erpemslg" = alloca i32
  %"tmp_clisp.inp_1-szoccafe" = alloca i32
  %"tmp_clisp.inp_0-doqcfdlm" = alloca i32
  %"tmp_clisp.inp_1-edumjpkf" = alloca i32
  %"tmp_clisp-tgleieml" = alloca i32
  %"tmp_clisp-nrujdajx" = alloca float*
  %"tmp_clisp.inp_0-hginnuwd" = alloca float
  %"tmp_clisp-tmtyncui" = alloca i32
  %"tmp_clisp-haukttod" = alloca float*
  %"tmp_clisp.inp_1-ouflrbka" = alloca float
  %"tmp_clisp.inp_1-rkturhmt" = alloca float
  %"tmp_clisp-hovlhhsi" = alloca float
  %"tmp_clisp.inp_0-jkjsczxt" = alloca i32
  %"tmp_clisp.inp_1-juxjtpcq" = alloca i32
  %"tmp_clisp-imybfwyj" = alloca i32
  %"tmp_clisp-cejghyey" = alloca i32
  %"tmp_clisp.ptr-bgtckjbj" = alloca float*
  %"tmp_clisp.val-zhylzszl" = alloca float
  %"tmp_clisp-kgfuitee" = alloca float
  %"tmp_clisp.inp_0-ckkbtnlq" = alloca i32
  %"tmp_clisp.inp_1-kfxbxrxh" = alloca i32
  %"tmp_clisp-rrkrmwxv" = alloca i32
  br label %"entry-ojeudlal"
entry-ojeudlal:
  store float* %"xout", float** %"xout.1"
  store float* %"x", float** %"x.1"
  store float* %"w", float** %"w.1"
  store i32 %"n", i32* %"n.1"
  store i32 %"d", i32* %"d.1"
  store i32 0, i32* %"tmp_clisp-myvvztwi"
  %".13" = load i32, i32* %"tmp_clisp-myvvztwi"
  store i32 %".13", i32* %"i"
  br label %"tmp_clisp.loop-wqclvjzg"
tmp_clisp.loop-wqclvjzg:
  %".16" = load i32, i32* %"i"
  store i32 %".16", i32* %"tmp_clisp.inp_0-tifpaiid"
  %".18" = load i32, i32* %"d.1"
  store i32 %".18", i32* %"tmp_clisp.inp_1-qqagurna"
  %".20" = load i32, i32* %"tmp_clisp.inp_0-tifpaiid"
  %".21" = load i32, i32* %"tmp_clisp.inp_1-qqagurna"
  %"tmp_clisp.cond-yoeqgrzo.1" = icmp slt i32 %".20", %".21"
  store i1 %"tmp_clisp.cond-yoeqgrzo.1", i1* %"tmp_clisp.cond-yoeqgrzo"
  %".23" = load i1, i1* %"tmp_clisp.cond-yoeqgrzo"
  br i1 %".23", label %"tmp_clisp.cont-zllyoxwk", label %"tmp_clisp.break-aphfxdtk"
tmp_clisp.cont-zllyoxwk:
  store float              0x0, float* %"tmp_clisp-yzjtqbcz"
  %".26" = load float, float* %"tmp_clisp-yzjtqbcz"
  store float %".26", float* %"val.twcwjsjw"
  store i32 0, i32* %"tmp_clisp-eqmzlcai"
  %".29" = load i32, i32* %"tmp_clisp-eqmzlcai"
  store i32 %".29", i32* %"j.twcwjsjw"
  br label %"tmp_clisp.loop-mrrebthr"
tmp_clisp.loop-mrrebthr:
  %".32" = load i32, i32* %"j.twcwjsjw"
  store i32 %".32", i32* %"tmp_clisp.inp_0-lgfovtpl"
  %".34" = load i32, i32* %"n.1"
  store i32 %".34", i32* %"tmp_clisp.inp_1-bguarjjv"
  %".36" = load i32, i32* %"tmp_clisp.inp_0-lgfovtpl"
  %".37" = load i32, i32* %"tmp_clisp.inp_1-bguarjjv"
  %"tmp_clisp.cond-prqqdmwu.1" = icmp slt i32 %".36", %".37"
  store i1 %"tmp_clisp.cond-prqqdmwu.1", i1* %"tmp_clisp.cond-prqqdmwu"
  %".39" = load i1, i1* %"tmp_clisp.cond-prqqdmwu"
  br i1 %".39", label %"tmp_clisp.cont-mlgfkxdw", label %"tmp_clisp.break-hbyfzxit"
tmp_clisp.cont-mlgfkxdw:
  %".41" = load float, float* %"val.twcwjsjw"
  store float %".41", float* %"tmp_clisp.inp_0-xvbgbjas"
  %".43" = load i32, i32* %"i"
  store i32 %".43", i32* %"tmp_clisp.inp_0-erpemslg"
  %".45" = load i32, i32* %"n.1"
  store i32 %".45", i32* %"tmp_clisp.inp_1-szoccafe"
  %".47" = load i32, i32* %"tmp_clisp.inp_0-erpemslg"
  %".48" = load i32, i32* %"tmp_clisp.inp_1-szoccafe"
  %"tmp_clisp.inp_0-doqcfdlm.1" = mul i32 %".47", %".48"
  store i32 %"tmp_clisp.inp_0-doqcfdlm.1", i32* %"tmp_clisp.inp_0-doqcfdlm"
  %".50" = load i32, i32* %"j.twcwjsjw"
  store i32 %".50", i32* %"tmp_clisp.inp_1-edumjpkf"
  %".52" = load i32, i32* %"tmp_clisp.inp_0-doqcfdlm"
  %".53" = load i32, i32* %"tmp_clisp.inp_1-edumjpkf"
  %"tmp_clisp-tgleieml.1" = add i32 %".52", %".53"
  store i32 %"tmp_clisp-tgleieml.1", i32* %"tmp_clisp-tgleieml"
  %".55" = load float*, float** %"w.1"
  %".56" = load i32, i32* %"tmp_clisp-tgleieml"
  %".57" = getelementptr float, float* %".55", i32 %".56"
  store float* %".57", float** %"tmp_clisp-nrujdajx"
  %".59" = load float*, float** %"tmp_clisp-nrujdajx"
  %".60" = load float, float* %".59"
  store float %".60", float* %"tmp_clisp.inp_0-hginnuwd"
  %".62" = load i32, i32* %"j.twcwjsjw"
  store i32 %".62", i32* %"tmp_clisp-tmtyncui"
  %".64" = load float*, float** %"x.1"
  %".65" = load i32, i32* %"tmp_clisp-tmtyncui"
  %".66" = getelementptr float, float* %".64", i32 %".65"
  store float* %".66", float** %"tmp_clisp-haukttod"
  %".68" = load float*, float** %"tmp_clisp-haukttod"
  %".69" = load float, float* %".68"
  store float %".69", float* %"tmp_clisp.inp_1-ouflrbka"
  %".71" = load float, float* %"tmp_clisp.inp_0-hginnuwd"
  %".72" = load float, float* %"tmp_clisp.inp_1-ouflrbka"
  %"tmp_clisp.inp_1-rkturhmt.1" = fmul float %".71", %".72"
  store float %"tmp_clisp.inp_1-rkturhmt.1", float* %"tmp_clisp.inp_1-rkturhmt"
  %".74" = load float, float* %"tmp_clisp.inp_0-xvbgbjas"
  %".75" = load float, float* %"tmp_clisp.inp_1-rkturhmt"
  %"tmp_clisp-hovlhhsi.1" = fadd float %".74", %".75"
  store float %"tmp_clisp-hovlhhsi.1", float* %"tmp_clisp-hovlhhsi"
  %".77" = load float, float* %"tmp_clisp-hovlhhsi"
  store float %".77", float* %"val.twcwjsjw"
  %".79" = load i32, i32* %"j.twcwjsjw"
  store i32 %".79", i32* %"tmp_clisp.inp_0-jkjsczxt"
  store i32 1, i32* %"tmp_clisp.inp_1-juxjtpcq"
  %".82" = load i32, i32* %"tmp_clisp.inp_0-jkjsczxt"
  %".83" = load i32, i32* %"tmp_clisp.inp_1-juxjtpcq"
  %"tmp_clisp-imybfwyj.1" = add i32 %".82", %".83"
  store i32 %"tmp_clisp-imybfwyj.1", i32* %"tmp_clisp-imybfwyj"
  %".85" = load i32, i32* %"tmp_clisp-imybfwyj"
  store i32 %".85", i32* %"j.twcwjsjw"
  br label %"tmp_clisp.loop-mrrebthr"
tmp_clisp.break-hbyfzxit:
  %".88" = load i32, i32* %"i"
  store i32 %".88", i32* %"tmp_clisp-cejghyey"
  %".90" = load float*, float** %"xout.1"
  %".91" = load i32, i32* %"tmp_clisp-cejghyey"
  %".92" = getelementptr float, float* %".90", i32 %".91"
  store float* %".92", float** %"tmp_clisp.ptr-bgtckjbj"
  %".94" = load float, float* %"val.twcwjsjw"
  store float %".94", float* %"tmp_clisp.val-zhylzszl"
  %".96" = load float*, float** %"tmp_clisp.ptr-bgtckjbj"
  %".97" = load float, float* %"tmp_clisp.val-zhylzszl"
  store float %".97", float* %".96"
  %".99" = load float, float* %"tmp_clisp.val-zhylzszl"
  store float %".99", float* %"tmp_clisp-kgfuitee"
  %".101" = load i32, i32* %"i"
  store i32 %".101", i32* %"tmp_clisp.inp_0-ckkbtnlq"
  store i32 1, i32* %"tmp_clisp.inp_1-kfxbxrxh"
  %".104" = load i32, i32* %"tmp_clisp.inp_0-ckkbtnlq"
  %".105" = load i32, i32* %"tmp_clisp.inp_1-kfxbxrxh"
  %"tmp_clisp-rrkrmwxv.1" = add i32 %".104", %".105"
  store i32 %"tmp_clisp-rrkrmwxv.1", i32* %"tmp_clisp-rrkrmwxv"
  %".107" = load i32, i32* %"tmp_clisp-rrkrmwxv"
  store i32 %".107", i32* %"i"
  br label %"tmp_clisp.loop-wqclvjzg"
tmp_clisp.break-aphfxdtk:
}
```

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
The new approach to return statements ensures that the last basic block is always terminated. The last label in the emitted Brilisp holds a return statement that returns a certain return variable, the name of which is fixed for a function. Codegen for `ret` in C-Lisp now emits Brilisp instructions to set the function's return variable and jump to the last label.

Additionally, implicit returns are now detected and reported as errors.
EDIT: We are letting these pass without attempting to detect them, at least in the C-Lisp layer. If a control path doesn't contain a return statement, traversing that path will result in an junk return value

Fixes #38

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
An error is now reported when an undeclared function is called.

**Comment by GlowingScrewdriver** [comment]
This is related to #31

**Comment by chsasank** [comment]
Add a regression test from #38 and add more tests

File: src/backend/c-lisp.py [hunk 1]
```diff
@@ -300,6 +318,8 @@ def gen_call_expr(self, expr, res_sym):
             arg_syms.append(arg_sym)
             instr_list += self.gen_expr(arg, res_sym=arg_sym)
         name = expr[1]
+        if not name in self.function_types:
+            raise CodegenError(f"Call to undeclared function: {name}")
```
Comment (chsasank) [inline src/backend/c-lisp.py:322]: Add test to verify that you're indeed getting an error

**Comment by GlowingScrewdriver** [comment]
> Add a regression test from #38 and add more tests

I think we should just enforce the rule that the last statement in the function should be a return statement. Because the alternative is to analyze the code and determine whether the function returns in all control paths or not -- that is complicated at best, and impossible at worst. Consider this example:
```scheme
(define ((find int) (arr (ptr int)) (n int) (key (int)))
    (declare (i int))
    (for ((set i 0)
            (lt i n)
            (set i (add i 1)))
        (if (eq (load (ptradd arr i)) key)
            (ret i))))
```
It is not guaranteed that the return statement inside the for loop (`ret i`) is reached.
When I compile the equivalent C code using Clang:
```C
#include <stdio.h>

int fn (int* arr, int n, int k) {
    for (int i=0; i<n; i++)
        if (arr[i] == k)
            return i;
}

int main () {
    int a [] = {1, 2, 3, 4, 5};
    printf ("%d\n", fn (a, 5, 6));
    return 0;
}
```
Clang emits a warning:
```
warning: non-void function does not return a value in all control paths [-Wreturn-type]
```
And the resulting code prints a random value when run, because the return variable is allocated and returned without being assigned.

Maybe we can allow just void functions to be without a return statement.

**Comment by GlowingScrewdriver** [comment]
Test infrastrure for errors is now up, at [src/backend/tests/c-lisp/errors](src/backend/tests/c-lisp/errors).

File: src/backend/c-lisp.py [hunk 2]
```diff
@@ -115,7 +148,7 @@ def gen_stmt(self, stmt):
             else:
                 return self.gen_expr(stmt)
         except Exception as e:
-            print(f"Error in statement: {stmt}")
+            print(f"Error in statement: {stmt}", file=sys.stderr)
```
Comment (chsasank) [inline src/backend/c-lisp.py:151]: no prints

File: src/backend/c-lisp.py [hunk 3]
```diff
@@ -87,10 +87,43 @@ def gen_function(self, func):
...
+            self.ret_ptr_sym, ret_label, ret_alloc_size_sym, ret_val_sym = [
+                random_label(CLISP_PREFIX, [extra])
+                for extra in ("ret_ptr", "ret_lbl", "ret_alloc", "ret_val")
+            ]
+            self.ret_jmp_instr = ["jmp", ret_label]
+            if ret_type == "void":
+                ret_alloc_instrs = []
+                ret_instrs = [["ret"]]
+            else:
+                ret_alloc_instrs = [
```
Comment (chsasank) [inline src/backend/c-lisp.py:102]: document bit more

File: src/backend/c-lisp.py [hunk 4]
```diff
@@ -87,10 +87,51 @@ def gen_function(self, func):
...
+
+        # If the function has a body, we need to codegen for it
+        if len(func) > 2:
+            # Set up a return label and a return variable, if needed
+            self.ret_ptr_sym, ret_label, ret_alloc_size_sym, ret_val_sym = [
+                random_label(CLISP_PREFIX, [extra])
+                for extra in ("ret_ptr", "ret_lbl", "ret_alloc", "ret_val")
+            ]
+            self.ret_jmp_instr = ["jmp", ret_label]
+            if not ret_type == "void":
```
Comment (chsasank) [inline src/backend/c-lisp.py:99]: use` ret_type != "void"`

**Comment by chsasank** [comment]
Add lot more tests. Look up all the codegen errors in c-lisp.py and ensure there are 2 tests per each instance.

File: src/backend/c-lisp.py [hunk 5]
```diff
@@ -300,6 +341,8 @@ def gen_call_expr(self, expr, res_sym):
             arg_syms.append(arg_sym)
             instr_list += self.gen_expr(arg, res_sym=arg_sym)
         name = expr[1]
+        if not name in self.function_types:
```
Comment (chsasank) [inline src/backend/c-lisp.py:344]: `name not in self.function_types` is more pythonic

File: src/backend/c-lisp.py (hunk 4 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

File: src/backend/c-lisp.py (hunk 5 above)
Comment (GlowingScrewdriver) [reply src/backend/c-lisp.py]: Done

**Comment by GlowingScrewdriver** [comment]
function `verify_shape` (in [utils/shapes.py](src/backend/utils/shapes.py)) has been implemented to enable more robust syntax error detection. The function can recursively verify the length and member types of any list; for example, the lists
```
["set", "var", 5]
["set", "var", ["add", "var", 1]]
```
can be verified against a pattern represented as
```
[str, str, None]
```

`verify_shape` will also serve the purpose of replacing the extensive (yet insufficient) pattern of checking list lengths in c-lisp.py. 
Check out the [codegen for `gen_set_expr`](https://github.com/chsasank/llama.lisp/pull/49/files#diff-70dacaf63181e637ba331a8fc351a098abb0f1a62bd7f1042548391431cf826eR312) to see what the refactor will look like.

**Comment by chsasank** [comment]
I have seen `verify_shape` and I think it's worth making up a separate PR and discussion on it.

**Comment by chsasank** [comment]
Can you confirm all the tests pass, I'll merge?

**Comment by GlowingScrewdriver** [comment]
> I have seen `verify_shape` and I think it's worth making up a separate PR and discussion on it.
Sure. I'll revert it on this branch

> Can you confirm all the tests pass, I'll merge?

Yes. Tests are passing

**Comment by chsasank** [comment]
Leave it on utils. We'll revisit later. Create an issue for python based pattern matching. We can use this as reference: https://github.com/chsasank/llama.lisp/blob/main/src/pat-match.lisp

**Pull request merged**